waspc/src/Wasp/Generator/Valid/PackageJson.hs

```diff
@@ -3,172 +3,162 @@ module Wasp.Generator.Valid.PackageJson
   where
-    validateRequiredWorkspaces Nothing = [missingWorkspacesError]
-    validateRequiredWorkspaces (Just definedWorkspacesList)
-      | NW.workspaceGlobs `S.isSubsetOf` definedWorkspaces = []
-      | otherwise = [makeWrongWorkspacesError definedWorkspaces]
-      where
-        definedWorkspaces = S.fromList definedWorkspacesList
-
-    makeWrongWorkspacesError definedWorkspaces =
-      GenericGeneratorError $
-        unwords
-          [ "Wasp requires package.json \"workspaces\" to include:",
-            showSet NW.workspaceGlobs ++ ".",
-            "You are missing:",
-            showSet (NW.workspaceGlobs `S.difference` definedWorkspaces) ++ "."
-          ]
-
-    missingWorkspacesError =
-      GenericGeneratorError $
-        unwords
-          [ "Wasp requires package.json \"workspaces\" to be present with the value and include values:",
-            showSet NW.workspaceGlobs ++ "."
-          ]
-
-    showSet = intercalate ", " . fmap show . S.toList
-
-validatePackageJsonDependency :: P.PackageJson -> PackageSpecification -> PackageRequirement -> [GeneratorError]
-validatePackageJsonDependency packageJson (packageName, expectedPackageVersion) requirement =
-  case maybePackageJsonDepedency of
-    Just actualPackageVersion ->
-      if actualPackageVersion == expectedPackageVersion
-        then []
-        else [incorrectPackageVersionErrorMessage]
-    Nothing ->
-      if isInWrongLocation requirement
-        then [wrongDependencyTypeErrorMessage]
-        else getMissingPackageError requirement
+    checkVersion = eqVersion pkgVersion
+
+    inCorrectDepList = inDependency depType dep
+    inOppositeDepList = inDependency (oppositeDepType depType) dep
+
+    oppositeDepType Runtime = Development
+    oppositeDepType Development = Runtime
+
+    resultForOpposite NotPresent = pure ()
+    resultForOpposite _ = wrongDepTypeError
+
+    resultForCorrect CorrectVersion = pure ()
+    resultForCorrect NotPresent = missingPackageError
+    resultForCorrect IncorrectVersion = incorrectPackageVersionError
+
+    wrongDepTypeError =
+      failure $
+        "Wasp requires package " ++ show pkgName ++ " to be in " ++ show (depTypeFieldName depType) ++ "."
+
+    missingPackageError =
+      failure $
+        "Wasp requires package " ++ show pkgName ++ " with version " ++ show pkgVersion ++ "."
+
+    incorrectPackageVersionError =
+      failure $
+        "Wasp requires package " ++ show pkgName ++ " to be version " ++ show pkgVersion ++ "."
+
+type PackageSpecification = (P.PackageName, P.PackageVersion)
+
+-- | Runs the validator on a specific dependency of the input, setting the appropriate path for
+-- errors.
+inDependency ::
+  DependencyType -> PackageSpecification -> Validator (Maybe P.PackageVersion) a -> Validator P.PackageJson a
+inDependency depType (pkgName, _) innerValidator =
+  inField (depTypeFieldName depType) (depTypeGetter depType) $
+    inField pkgName (M.lookup pkgName) innerValidator
```

Comment:
This `innerValidator`, that runs on what, on the version string? Is there a name for this `inner` thing, that would make it clearer what it is running on?